# Atividade T1.3 (Correto)

A "spline" cúbica dos pontos (−2,4), (0,1) e (1,−5) é uma função 

$$
f(x)={ax^{3}+bx^{2}+cx+d, se −2≤x≤0}\\
f(x)=ex^{3}+fx^{2}+gx+h, se 0≤x≤1
$$

Determine os valores de a,b,c,d,e,f,g na forma uma fração em que o denominador é sempre positivo e, no caso de uma fração inteira, igual a 1.

Resolução:

Primeiro vamos calcular o número de equações que precisamos extrair:

$$
n+1=3\\
n=2
$$

Logo, precisamos de $4*2=8$ equações. Nós temos:
$$
x_0=-2, y_0=4\\
x_1=0, y_1=1\\
x_2=1, y_2=-5
$$

A seguir temos 4 propriedades que extraem equações da "s-pline":

### Propriedade 1 (2 equações)

Nos extremos $x_0 $ e $ x_n$:

$$ 
P_1(x_0)=y_0\\
P_n(x_n)=y_n\\
\Downarrow\\
P_1(-2) = 4 \Rightarrow a(-2)^{3} + b(-2)^{2} + c(-2) + d = 4\\
P_2(1) = -5 \Rightarrow e(1)^{3} + f(1)^{2} + g(1) + h = -5\\
\Downarrow\\
-8a+4b-2c+d=4\\
e+f+g+h=-5
$$


### Propriedade 2 (2(n-1) equações)

Continuidade da $f$ impõe:

$$
P_1(x_1) = y_1=P_2(x_1)\\
P_2(x_2) = y_2=P_3(x_2)\\
...\\
P_{n-1}(x_{n-1}) = y_{n-1}=P_n(x_{n-1})\\
$$

Logo,

$$
P_1(0) = 1 \Rightarrow d = 1\\
P_2(0) = 1 \Rightarrow h = 1
$$

### Propriedade 3 (n+1 equações)

Derivadas de 1º ordem:

$$
P'_1(x_0)=0\\
P'_n(x_n)=0\\
P'_1(x_1)=P'_2(x_1)\\
...\\
P'_{n-1}(x_{n-1})=P'_n(x_{n-1})
$$

Logo,

$$
P'_1(x_0)=0 \Rightarrow 3a(-2)^{2}+2b(-2)+c=0\\
P'_2(x_2)=0 \Rightarrow 3e(1)^{2}+2f(1)+g=0\\
P'_1(x_1)=P'_2(x_1) \Rightarrow c=g\\
\Downarrow\\
12a-4b+c=0\\
3e+2f+g=0\\
c=g
$$


### Propriedade 4 (n-1 equações)

Derivadas de 2º ordem contínuas:

$$
P''_1(x_1)=P''_2(x_1)\\
...\\
P''_{n-1}(x_{n-1})=P''_n(x_{n-1})\\
$$

Logo,

$$
P''_1(x_1)=P''_2(x_1) \Rightarrow 2b=2f\\
\Downarrow\\
2b=2f
$$


Assim, temos o sistema linear (nas incógnitas a,b,c,d,e,f,g,h):

$$
\left\{
\begin{matrix}
-8a &+&  4b &-&  2c &+& d &+& 0e &+& 0f &+& 0g &+& 0h = 4 \\
0a &+&  0b &+&  0c &+& 0d &+& e &+& f &+& g &+& h = -5\\
0a &+&  0b &+&  0c &+& d &+& 0e &+& 0f &+& 0g &+& 0h = 1\\ 
0a &+&  0b &+&  0c &+& 0d &+& 0e &+& 0f &+& 0g &+& h = 1\\
12a &-&  4b &+&  c &+& 0d &+& 0e &+& 0f &+& 0g &+& 0h = 0\\
0a &+&  0b &+&  0c &+& 0d &+& 3e &+& 2f &+& g &+& 0h = 0\\
0a &+&  0b &+&  c &+& 0d &+& 0e &+& 0f &-& g &+& 0h = 0\\
0a &+&  2b &+&  0c &+& 0d &+& 0e &-& 2f &+& 0g &+& 0h = 0\\
\end{matrix}
\right\rvert\\
\Downarrow\\
\begin{bmatrix}
-8 &  4 & -2 & 1 & 0 & 0 & 0 & 0 & 4 \\
0 &  0 &  0 & 0 & 1 & 1 & 1 & 1 & -5\\
0 &  0 &  0 & 1 & 0 & 0 & 0 & 0 & 1\\ 
0 &  0 &  0 & 0 & 0 & 0 & 0 & 1 & 1\\
12 & -4 & 1  & 0 & 0 & 0 & 0 & 0 & 0\\
0 &  0 &  0 & 0 & 3 & 2 & 1 & 0 & 0\\
0 &  0 &  1 & 0 & 0 & 0 & -1 & 0 & 0\\
0 &  2 &  0 & 0 & 0 & -2 & 0 & 0 & 0
\end{bmatrix}
$$

In [2]:
A = [-8 ,  4 , -2 , 1 , 0 , 0 , 0 , 0;
0 ,  0 ,  0 , 0 , 1 , 1 , 1 , 1;
0 ,  0 ,  0 , 1 , 0 , 0 , 0 , 0; 
0 ,  0 ,  0 , 0 , 0 , 0 , 0 , 1;
12 , -4 , 1  , 0 , 0 , 0 , 0 , 0;
0 ,  0 ,  0 , 0 , 3 , 2 , 1 , 0;
0 ,  0 ,  1 , 0 , 0 , 0 , -1 , 0;
0 ,  2 ,  0 , 0 , 0 , -2 , 0 , 0]
b=[4; -5; 1; 1; 0; 0; 0 ; 0]

[x,A,b] = gaussPivotP(A,b)

A =

   -8    4   -2    1    0    0    0    0
    0    0    0    0    1    1    1    1
    0    0    0    1    0    0    0    0
    0    0    0    0    0    0    0    1
   12   -4    1    0    0    0    0    0
    0    0    0    0    3    2    1    0
    0    0    1    0    0    0   -1    0
    0    2    0    0    0   -2    0    0

b =

   4
  -5
   1
   1
   0
   0
   0
   0

x =

  -0.9375
  -4.5000
  -6.7500
   1.0000
   5.2500
  -4.5000
  -6.7500
   1.0000

A =

   12.0000   -4.0000    1.0000         0         0         0         0         0
         0    2.0000         0         0         0   -2.0000         0         0
         0         0   -1.3333    1.0000         0    1.3333         0         0
         0         0         0    1.0000         0         0         0         0
         0         0         0         0    3.0000    2.0000    1.0000         0
         0         0         0         0         0    1.0000   -1.0000         0
         0         0         0         0   

### Resultado

Chegamos no vetor solução:

$$
\begin{bmatrix}
  -0.9375\\
  -4.5000\\
  -6.7500\\
   1.0000\\
   5.2500\\
  -4.5000\\
  -6.7500\\
   1.0000
\end{bmatrix}
=

\begin{bmatrix}
-\frac{15}{16}\\
-\frac{9}{2}\\
-\frac{27}{4}\\
1\\
\frac{21}{4}\\
-\frac{9}{2}\\
-\frac{27}{4}\\
1
\end{bmatrix}
$$
